src/sgame/components/SpikerComponent.cpp

```diff
@@ -19,175 +39,215 @@ void SpikerComponent::HandleDamage(float amount, gentity_t *source, Util::option
+	float missileSize = (float)ma->size;
+	trace_t trace;
+	vec3_t mins, maxs;
+	Vec3 end = Vec3::Load(entity.oldEnt->s.origin) + (SPIKE_RANGE * direction);
+
+	// Test once with normal and once with inflated missile bounding box.
+	for (float traceSize : {missileSize, missileSize * SAFETY_TRACE_INFLATION}) {
+		mins[0] = mins[1] = mins[2] = -traceSize;
+		maxs[0] = maxs[1] = maxs[2] =  traceSize;
+		trap_Trace(&trace, entity.oldEnt->s.origin, mins, maxs, end.Data(), entity.oldEnt->s.number,
+			ma->clipmask, 0);
+		gentity_t* hit = &g_entities[trace.entityNum];
+
+		if (hit && G_OnSameTeam(entity.oldEnt, hit)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
 bool SpikerComponent::Fire() {
 	gentity_t *self = entity.oldEnt;
 	// Check if still resting.
 	if (restUntil > level.time) {
+		logger.Verbose("Spiker #%i wanted to fire but wasn't ready.", entity.oldEnt->s.number);
+
 		return false;
+	} else {
+		logger.Verbose("Spiker #%i is firing!", entity.oldEnt->s.number);
 	}
 
 	// Play shooting animation.
 	G_SetBuildableAnim(self, BANIM_ATTACK1, false);
+	GetBuildableComponent().ProtectAnimation(5000);
 
 	// TODO: Add a particle effect.
 	//G_AddEvent(self, EV_ALIEN_SPIKER, DirToByte(self->s.origin2));
 
 	// Calculate total perimeter of all spike rows to allow for a more even spike distribution.
 	float totalPerimeter = 0.0f;
 
-	for (int row = 0; row < SPIKER_MISSILEROWS; row++) {
-		float altitude = (((float)row + SPIKER_ROWOFFSET) * M_PI_2) / (float)SPIKER_MISSILEROWS;
+	for (int row = 0; row < MISSILEROWS; row++) {
+		float altitude = (((float)row + ROWOFFSET) * M_PI_2) / (float)MISSILEROWS;
 		float perimeter = 2.0f * M_PI * cos(altitude);
 
 		totalPerimeter += perimeter;
 	}
 
 	// Distribute and launch missiles.
+	// TODO: Use new vector library.
 	vec3_t dir, rowBase, zenith, rotAxis;
 
 	VectorCopy(self->s.origin2, zenith);
 	PerpendicularVector(rotAxis, zenith);
 
-	for (int row = 0; row < SPIKER_MISSILEROWS; row++) {
-		float altitude = (((float)row + SPIKER_ROWOFFSET) * M_PI_2) / (float)SPIKER_MISSILEROWS;
+	for (int row = 0; row < MISSILEROWS; row++) {
+		float altitude = (((float)row + ROWOFFSET) * M_PI_2) / (float)MISSILEROWS;
 		float perimeter = 2.0f * M_PI * cos(altitude);
 
 		RotatePointAroundVector(rowBase, rotAxis, zenith, RAD2DEG(M_PI_2 - altitude));
 
 		// Attempt to distribute spikes with equal distance on all rows.
-		int spikes = (int)round(((float)SPIKER_MISSILES * perimeter) / totalPerimeter);
+		int spikes = (int)round(((float)MISSILES * perimeter) / totalPerimeter);
 
 		for (int spike = 0; spike < spikes; spike++) {
 			float azimuth = 2.0f * M_PI * (((float)spike + 0.5f * crandom()) / (float)spikes);
-			float altitudeVariance = 0.5f * crandom() * M_PI_2 / (float)SPIKER_MISSILEROWS;
+			float altitudeVariance = 0.5f * crandom() * M_PI_2 / (float)MISSILEROWS;
 
 			RotatePointAroundVector(dir, zenith, rowBase, RAD2DEG(azimuth));
 			RotatePointAroundVector(dir, rotAxis, dir, RAD2DEG(altitudeVariance));
```

Comment:
This line doesn't look right. To change (only) the altitude, the rotation axis can't just be any vector perpendicular to `zenith`; it should be the cross product of `zenith` and `dir`. 
Maybe it would be easier to use a spherical coordinate system.

Reply:
Hmm, it should work if the two rotations are swapped, right?

What do you mean exactly by using a spherical coordinate system here? Isn't this already what I'm doing as I rotate the unit vector zenith first by the angle altitude and then by the angle perimeter?

Reply:
Yeah, I guess the only real difference would be to do the entire altitude rotation at once, which you have already done now.